### PR TITLE
Embed the topology in the NetworkBehaviour 

### LINKED ISSUE #854: Embed the topology directly in the NetworkBehaviour
cc #785 

Right now libp2p assumes one single network, and therefore one single topology.
In practice, though, we probably want to support multiple overlay networks, potentially on top of one another.

In these situations, having one single `Topology` plugged into the `Swarm` is not necessarily an ideal solution.

Instead I suggest that:

- Some implementations of `NetworkBehaviour` hold a topology (example: `KademliaDiscovery`, `Brahms`, `Mdns`).
- When asked to dial a peer, the swarm retrieves the dialing addresses by calling a method on `NetworkBehaviour`. Multiple behaviours can be plugged together by joining the results.
- The user can create complex layouts by "plugging" the multiple behaviours together manually.
- `Floodsub` would get passed a list of destinations every time we send a message (cc #713).

For example, a user can:

- Have a `HashMap<NetworkId, KademliaDiscovery>`, where each `KademliaDiscovery` holds a partial random view of its network.
- The `KademliaDiscovery` generates events that indicate when this view is updated.
- When they receive such an event, the user calls a method on the corresponding `Gossipsub` (for example) in a `HashMap<NetworkId, Gossipsub>`.
- `Gossipsub` handles building an overlay network of top of the view of the `KademliaDiscovery`.

The drawback is that the user now has to do things a bit more manually. For example when `identify` generates events containing the listening addresses of a node we're connected to, the user has to transmit this information to the `KademliaDiscovery`.

This solution would work around #785 by removing the problem altogether.

### PR DESCRIPTION
**Pull request by tomaka**
Close #854 
See the issue for description.
Also cc'ing #843 

Changes:

- Entirely removes the `topology` module and its content. You now have to pass a `PeerId` (id of the local node) instead of a `Topology` when creating the `Swarm`.
- The user now has to manually pass to `Floodsub` the list of peer IDs that should be considered for gossiping. See the changes in `chat.rs`. While this seems annoying, we should probably add network behaviours that automatically do that in the future.
- Kademlia and mDNS are now purely discovery protocols and now report the list of nodes that they discover.
- The implementation of k-buckets has changed. Normally in the Kademlia paper you're supposed to ping the oldest node when you discard it. In libp2p we need to build a connection in order to ping a node, therefore instead of "pinging the node" we just try to connect to it instead. If we are already connected to this node, we assume that it is responsive. Consequently, if a bucket is full of nodes we are already connected to, we refuse any new insertion.

### CODE REVIEW

File: core/src/swarm.rs [hunk 1]
```diff
@@ -139,29 +137,12 @@ where TBehaviour: NetworkBehaviour<TTopology>,
...
-        let raw_swarm = RawSwarm::new(transport, topology.local_peer_id().clone());
-
-        Swarm {
-            raw_swarm,
-            behaviour,
-            topology,
-            supported_protocols,
-            listened_addrs: SmallVec::new(),
-        }
+    pub fn new(transport: TTransport, behaviour: TBehaviour, local_peer_id: PeerId) -> Self {
```
Comment (twittner) [inline core/src/swarm.rs]: With #881 merged we should make `Swarm` generic over `PeerId` as well here.

File: core/src/swarm.rs [hunk 2]
```diff
@@ -355,6 +317,10 @@ pub trait NetworkBehaviour<TTopology> {
     /// Builds a new `ProtocolsHandler`.
     fn new_handler(&mut self) -> Self::ProtocolsHandler;
 
+    /// Addresses that this behaviour is aware of for this specific peer, and that may allow
+    /// reaching the peer.
+    fn addresses_of_peer(&self, peer_id: &PeerId) -> Vec<Multiaddr>;
```
Comment (twittner) [inline core/src/swarm.rs]: This should likewise be generic over peer IDs.

Comment (tomaka) [reply core/src/swarm.rs]: It's not as easy, because then we have to make the `NetworkBehaviour` trait generic over peer IDs as well.

Comment (twittner) [reply core/src/swarm.rs]: Yes, I know.

**Pull request merged**